### ISSUE TITLE
calling pause from parent container

### ISSUE THREAD

**Issue by PradhumnSingh** [volkov97/react-compound-timer#29]
Hi, is there any way to call pause function from parent container?

**Comment by sammdec** [comment]
Yeah I would also love to know how to do this as well

**Comment by malinantonsson** [comment]
Hey,
You can create a ref on the parent container `const timerRef = useRef(null);`
```
<ReactTimer
  onPause={onStop}
  ref={timerRef} />
```
You'll then have access to all the functions, ie `timerRef.current.pause();`

**Comment by tigersze** [comment]
> Hey,
> You can create a ref on the parent container `const timerRef = useRef(null);`
> 
> ```
> <ReactTimer
>   onPause={onStop}
>   ref={timerRef} />
> ```
> 
> You'll then have access to all the functions, ie `timerRef.current.pause();`
@malinantonsson would you mind explaining more? i do not how to do this.

**Comment by tigersze** [comment]
@malinantonsson would you mind explaining more? i do not how to do this.

> Hey,
> You can create a ref on the parent container `const timerRef = useRef(null);`
> 
> ```
> <ReactTimer
>   onPause={onStop}
>   ref={timerRef} />
> ```
> 
> You'll then have access to all the functions, ie `timerRef.current.pause();`

**Comment by tigersze** [comment]
@malinantonsson would you mind explaining more? i do not know how to do this.

> Hey,
> You can create a ref on the parent container `const timerRef = useRef(null);`
> 
> ```
> <ReactTimer
>   onPause={onStop}
>   ref={timerRef} />
> ```
> 
> You'll then have access to all the functions, ie `timerRef.current.pause();`

**Comment by malinantonsson** [comment]
@tigersze  ok, so you would do something like this:
```
import React, { useRef } from "react";
import ReactTimer from "react-compound-timer";

function YourComponent() {
   const timerRef = useRef(null);

   const onStop =() => {
    timerRef.current.stop();
   };
   
    return ( <ReactTimer
        onStop={onStop}
        ref={timerRef} />
    )
}
```

You can read more about refs https://reactjs.org/docs/refs-and-the-dom.html
Basically you create a reference to the Timer component and with that reference, you can access all the functions

**Comment by tigersze** [comment]
@malinantonsson Thank you very much. In fact, what i want to do is to call the start of the timer from another function (in other file, not from the parent). Is the method similar to yours?

**Comment by tigersze** [comment]
Also one more question. I am using the class but not function for my component. It do not allow me to put const inside the class. I don't know whether it is the problem from vscode.

**Comment by malinantonsson** [comment]
@tigersze yes, it should be `start` (see the main docs for this library for all the functions). In your case, you'd probably have to set the ref in the parent component and pass it down to both the children (the timer function & the other function). I recommend you read this https://reactjs.org/docs/composition-vs-inheritance.html as your issue is more about how to structure react components rather than an issue with this specific library. 

Also, this link https://reactjs.org/docs/components-and-props.html#function-and-class-components will help you understand the difference between class and functional components :)

**Comment by tigersze** [comment]
@malinantonsson Thanks. I have successfully created the reference in my time component (a child class). Now I want to call the class function, i.e. onStop in your example, in other functions. I know it is not related to this library problem. I have searched in the internet to your something like `new class().onStop`. However, it is no working.

> @tigersze ok, so you would do something like this:
> 
> ```
> import React, { useRef } from "react";
> import ReactTimer from "react-compound-timer";
> 
> function YourComponent() {
>    const timerRef = useRef(null);
> 
>    const onStop =() => {
>     timerRef.current.stop();
>    };
>    
>     return ( <ReactTimer
>         onStop={onStop}
>         ref={timerRef} />
>     )
> }
> ```
> 
> You can read more about refs https://reactjs.org/docs/refs-and-the-dom.html
> Basically you create a reference to the Timer component and with that reference, you can access all the functions

**Comment by tigersze** [comment]
This is my code:
Time.js (I have used the start button to test the reference and it works)
`import React from "react";
import Timer from "react-compound-timer";
import "./App.css";
class Time extends React.Component {
  constructor(props) {
    super(props);
    // create a ref to store the textInput DOM element
    this.onStart = React.createRef();
    this.startClick = this.startClick.bind(this);
    this.state = { start: true, alert: false };
  }
  startClick = () => {
    this.onStart.current.start();
  };
  render() {
    return (
      <Timer
        ref={this.onStart}
        startImmediately={false}
        onPause={() =>
          fetch(
            "http://hub.dev.objectblocks.cc/....",
            { method: "GET" }
          ).then(() => {
            this.setState({ start: false });
          })
        }
        checkpoints={[
          {
            time: 10000,
            callback: () =>
              fetch(
                "http://hub.dev.objectblocks.cc/...",
                { method: "GET" }
              ).then(() => {
                this.setState({ alert: true });
              }),
          },
        ]}
      >
        {({ pause, start }) => (
          <React.Fragment>
            <div className="timer">
              <Timer.Minutes /> M<span>&nbsp;</span>
              <Timer.Seconds /> s
            </div>
            <button className="button" onClick={pause}>
              Pause
            </button>
            <button onClick={this.startClick}>start</button>
            <div className="alert">
              {this.state.alert ? "You have ONE minutes left!" : null}
            </div>
          </React.Fragment>
        )}
      </Timer>
    );
  }
}
export default Time;`
And I want to call the startClick function in the Event.js:
`import { toast } from "react-toastify";
import { generateStore, EventActions } from "@drizzle/store";
import drizzleOptions from "./drizzleOptions";
import Time from "./Time";
const contractEventNotifier = (store) => (next) => (action) => {
  if (action.type === EventActions.EVENT_FIRED) {
    const message = action.event.returnValues._message;
    const display = `${message}`;
    fetch(
      "http://hub.dev.objectblocks.cc/...",
      { method: "GET" }
    );
    toast.success(display, { position: toast.POSITION.TOP_RIGHT });
    new Time().startClick();
  }
  return next(action);
};

const appMiddlewares = [contractEventNotifier];

const store = generateStore({
  drizzleOptions,
  appMiddlewares,
  disableReduxDevTools: false, // enable ReduxDevTools!
});

// Use the store with DrizzleProvider
export default store;`
In fact, this is a code related to the blockchain (drizzle). You can just ignore those code.

**Comment by malinantonsson** [comment]
@tigersze  I'm not sure I can help you with this, but if you set up a code pen or something I can take a look. But basically you need to think about how you structure your react components and pass your functions down as props.

**Comment by tigersze** [comment]
@samjbmason 
sure. The `time.js` (I have created the reference and test with the start button and it works)
https://codepen.io/tiger-tiger-sze/pen/YzwKMwy
The `Event.js` 
https://codepen.io/tiger-tiger-sze/pen/VweZNeL?editors=0010
I want to call the from the startClick function from Time in contractEventNotifier function. 
In fact, This is a blockchain app. You can just ignore those code. The app works fine and I can provide you the whole app if you want.

**Comment by tigersze** [comment]
The main problem is that i will do the blockchain transaction and start the timer. I don't know how to write the code. I just know that I need to call it in the Event.js

**Comment by tigersze** [comment]
@malinantonsson if you want the full code, you can give me your email and i can send you through email.

**Comment by malinantonsson** [comment]
@tigersze I had a look and I don't think I can help you with this, sorry, I've never used drizzle. I would try to ask the question in the repo for `drizzle` instead (or on StackOverflow).
